src/main/ApplicationImpl.cpp

```diff
@@ -133,6 +135,80 @@ ApplicationImpl::ApplicationImpl(VirtualClock& clock, Config const& cfg)
     }
 }
 
+static void
+maybeRebuildLedger(Application& app, bool applyBuckets)
+{
+    std::set<LedgerEntryType> toRebuild;
+    auto& ps = app.getPersistentState();
+    for (auto let : xdr::xdr_traits<LedgerEntryType>::enum_values())
+    {
+        LedgerEntryType t = static_cast<LedgerEntryType>(let);
+        if (ps.shouldRebuildForType(t))
+        {
+            toRebuild.emplace(t);
+        }
+    }
+    if (toRebuild.empty())
+    {
+        return;
+    }
+
+    if (!app.getConfig().MODE_USES_IN_MEMORY_LEDGER)
+    {
+        app.getDatabase().clearPreparedStatementCache();
+        soci::transaction tx(app.getDatabase().getSession());
+
+        for (auto let : toRebuild)
+        {
+            switch (let)
+            {
+            case ACCOUNT:
+                LOG_INFO(DEFAULT_LOG, "Dropping accounts");
+                app.getLedgerTxnRoot().dropAccounts();
+                break;
+            case TRUSTLINE:
+                LOG_INFO(DEFAULT_LOG, "Dropping trustlines");
+                app.getLedgerTxnRoot().dropTrustLines();
+                break;
+            case OFFER:
+                LOG_INFO(DEFAULT_LOG, "Dropping offers");
+                app.getLedgerTxnRoot().dropOffers();
+                break;
+            case DATA:
+                LOG_INFO(DEFAULT_LOG, "Dropping accountdata");
+                app.getLedgerTxnRoot().dropData();
+                break;
+            case CLAIMABLE_BALANCE:
+                LOG_INFO(DEFAULT_LOG, "Dropping claimablebalances");
+                app.getLedgerTxnRoot().dropClaimableBalances();
+                break;
+            default:
+                abort();
+            }
+        }
+
+        tx.commit();
+    }
+
+    // No transaction is needed. ApplyBucketsWork breaks the apply into many
+    // small chunks, each of which has its own transaction. If it fails at some
+    // point in the middle, then rebuildledger will not be cleared so this will.
```

Comment:
Incomplete sentence? "... so this will." I think this should read "... so this will re-run on next startup."

Reply:
Fixed, thanks.